contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+
+1. The controller may deserialize the values of the AnyStates representing their 
+target Object state and perform a deep, semantic equality test. Here all 
+differences that do not constitute a mutation to the target Object state are 
+disregarded during the equivalence test.
+    - As it is imperative, this method may be brittle and hard to maintain.
+1. Alternatively, controllers may use a strategic merge patch to ignore fields 
+whose deletion, addition, or mutation does not constitute a revision.
+    - This approach requires all values to be JSON encodings of Objects, but the 
+    declarative nature will likely make the code more maintainable and less 
+    error prone (The business logic for the patch and comparison only needs to 
+    be implemented once).
+    
+Note that, as both methods provide a test for deep, semantic equivalence, 
+controllers that implement one method are not bound to use the same method 
+in the future. 
+    
+### Target Object State Reconciliation
+There are three ways for a controller to reconcile a generated Object with the 
+declared target Object state.
+
+1. If the target Object state is [equivalent](#version-equivalence) to the 
+target Object state associated with the generated Object, the controller will 
+update the associated [version tracking information](#version-tracking).
+1. If the Object can be updated in place to reconcile its state with the 
+current target Object state, a controller may update the Object in place 
+provided that the associated version tracking information is updated as well.
+1. Otherwise, the controller must destroy the Object and recreate it from the 
+current target Object state.
+
+## Kubectl
+
+Modifications to kubectl to leverage controller history are an optional 
+extension. Users can trigger rolling updates and rollbacks by modifying their 
+manifests and using `kubectl apply`. Controllers will be able to detect 
+revisions to their target Object state and perform 
+[reconciliation](#target-object-state-reconciliation) as necessary.
+
+### Viewing History 
+As controllers can use strategic merge patching as a method for 
+[version equivalence](#version-equivalence) testing, so can clients use it as a 
+method for reconstructing specification type Objects at a particular revision. 
+We propose the following command.
+
+```bash
+> kubectl history <controller-kind> <controller-name>
+```
+
+In the above `controller-kind` indicates the Kind of the controller (e.g. 
+ReplicaSet, StatefulSet, DaemonSet) and `controller-name` indicates the `.Name` 
+of the request subject. We propose that the output of this command be a list 
+of the `.Names` of AnyStates that comprise the revision history of the 
+controller.
+
+In order to view the application of a particular revision to the controller we 
+propose the following additional flag.
+
+```bash
+> kubectl history <controller-kind> <controller-name> --revision <any-state-name>
```

Comment:
Current implementation of `--revision` asks for an `int64`. Users get revision number from `kubectl history`. Switching it to `<any-state-name>` will be a huge user experience change. Same as `kubectl rollout undo --to-revision`.